### PR TITLE
Cryp 129/add auditors to test harness

### PR DESCRIPTION
**Pull request by Arash-Afshar**
https://polymath.atlassian.net/browse/CRYP-129

Work left to be done:
- [x] Update the existing cli and add support for the auditors (this is what is causing the build errors)
- [x] Add the `audit` cli
- [ ] Remove TODOs from the diff

### CODE REVIEW

File: mercat/cli/common/src/harness.rs [hunk 1]
```diff
@@ -98,12 +117,27 @@ impl TryFrom<(u32, String)> for Transfer {
...
+                .to_lowercase()
+                .split(",")
+                .map(|name| Party::try_from((name, PartyKind::Auditor)))
+                .collect::<Result<_, _>>()?
+        };
+        let tx_name = if caps.get(11) == None {
+            None
+        } else {
+            Some(caps[11].trim().to_string())
+        };
```
Comment (fmiguelgarcia) [inline mercat/cli/common/src/harness.rs:134]: If you enable the **case-insensitive flag** `(?i)` in your regex expression, you can avoid the allocation of each `String` in `cap[9]` parse, and you could use directly `&str` (0 deep-copies).


```suggestion
        let auditors = caps.get(9).map( |c| {
            c.split(",")
                .map(|n| Party::try_from((n, PartyKind::Auditor)))
                .collect::<Result<_, _>>()
        }).unwrap_or(Ok(vec![]))?;
        
        let tx_name = caps.get(11).map(|c| c.as_str().trim().to_string());
```

File: mercat/cli/common/src/harness.rs [hunk 2]
```diff
@@ -121,42 +157,64 @@ impl TryFrom<(u32, String)> for Transfer {
...
+            r"^issue ([a-zA-Z0-9()]+) ([0-9]+) ([a-zA-Z0-9]+)( auditors)?( [a-zA-Z0-9,]+)?( tx_name)?( [a-zA-Z0-9-_]+)?$",
+        )
+        .map_err(|_| Error::RegexError {
+            reason: String::from("Failed to compile the Issue regex"),
         })?;
         let caps = re.captures(&segment).ok_or(Error::RegexError {
             reason: format!("Pattern did not match {}", segment),
         })?;
         let ticker = caps[3].to_string().to_uppercase();
+        let auditors = if caps.get(5) == None {
```
Comment (fmiguelgarcia) [inline mercat/cli/common/src/harness.rs:191]: ditto

File: mercat/cli/common/src/harness.rs [hunk 3]
```diff
@@ -345,6 +455,12 @@ impl Transfer {
         let tx_id = self.tx_id;
         let reject = !self.mediator_approves;
         let cheat = self.mediator.cheater;
+        let auditors: Vec<String> = self
+            .auditors
+            .clone()
+            .into_iter()
+            .map(|auditor| auditor.name)
+            .collect();
```
Comment (fmiguelgarcia) [inline mercat/cli/common/src/harness.rs:463]: If you only need the `auditor.name`, only clone this member. 
Pls note that `into_iter` has been transformed into just `iter`, to use references (you don't need to consume it).
```suggestion
        let auditors = self.auditors.iter().map(|auditor| auditor.name.clone()).collect::<Vec<_>>();
```

File: mercat/cli/common/src/harness.rs [hunk 4]
```diff
@@ -460,18 +609,27 @@ impl Issue {
         let issuer = self.issuer.name.clone();
         let amount = self.amount;
         let tx_id = self.tx_id;
+        let tx_name = self.tx_name.clone();
         let cheat = self.issuer.cheater;
+        let auditors: Vec<String> = self
```
Comment (fmiguelgarcia) [inline mercat/cli/common/src/harness.rs:614]: ditto

File: mercat/cli/common/src/harness.rs [hunk 5]
```diff
@@ -668,23 +892,54 @@ fn all_dirs_in_dir(dir: PathBuf) -> Result<Vec<PathBuf>, Error> {
...
+}
+
+fn make_empty_non_normal_accounts(
+    accounts: &[(Party, Option<u32>)],
+    starting_id: u32,
+) -> Result<(u32, Vec<TransactionMode>), Error> {
+    let mut transaction_counter = starting_id;
+    let mut seq: Vec<TransactionMode> = vec![];
+    for account in accounts {
+        let (party, owner_id) = account;
```
Comment (fmiguelgarcia) [inline mercat/cli/common/src/harness.rs:927]: One step
```suggestion
    for (party, owner_id) in accounts {
```

**Review by fmiguelgarcia — COMMENTED** [review]
Some comments

File: mercat/cli/auditor/src/input.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    )]
+    pub save_config: Option<PathBuf>,
+
+    /// Instructs the CLI to act as a cheater.
+    #[structopt(long, help = "Instructs the CLI to act as a cheater.")]
+    pub cheat: bool,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, StructOpt)]
+pub struct AuditIransferInfo {
```
[the remark this answers is not in the document]
Comment (Arash-Afshar) [reply mercat/cli/auditor/src/input.rs:125]: Those were copied from other CLIs and are removed in this PR. But your comment is valid and we should do that in future refactorings.

File: mercat/cli/common/src/lib.rs [hunk 7]
```diff
@@ -98,25 +103,29 @@ impl CoreTransaction {
                 tx_id: _,
                 ordering_state: _,
                 amount: _,
+                auditors: _,
             } => true,
             CoreTransaction::TransferJustify {
                 tx: _,
                 mediator: _,
                 tx_id: _,
+                auditors: _,
```
[the remark this answers is not in the document]
Comment (Arash-Afshar) [reply mercat/cli/common/src/lib.rs:112]: Nice!

File: mercat/cli/common/src/harness.rs (hunk 1 above)
Comment (Arash-Afshar) [reply mercat/cli/common/src/harness.rs:134]: This did not work, we can chat on Monday.

File: mercat/cli/auditor/src/input.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    pub auditor: String,
+
+    /// An optional seed, to feed to the RNG, that can be passed to reproduce a previous run of this CLI.
+    /// The seed can be found inside the logs.
+    #[structopt(
+        short,
+        long,
+        help = "Base64 encoding of an initial seed for the RNG. If not provided, the seed will be chosen at random."
+    )]
+    pub seed: Option<String>,
```
[the remark this answers is not in the document]
Comment (Arash-Afshar) [reply mercat/cli/auditor/src/input.rs]: Will probably address in another refactoring PR.

File: mercat/cli/common/src/validate.rs [hunk 9]
```diff
@@ -276,6 +291,25 @@ pub fn validate_asset_issuance(
...
+            );
+
+            key
+        })
+        .collect::<Result<Vec<(u32, EncryptionPubKey)>, _>>()
+        .map_err(|error| {
+            error!("Error in validation of tx-{}: {:#?}", tx_id, error);
+            ValidationResult::error("user", "ticker")
+        });
+    let auditors = auditors.unwrap(); // TODO
```
[the remark this answers is not in the document]
Comment (Arash-Afshar) [reply mercat/cli/common/src/validate.rs]: I'll need to your help with that part of the code, the data structures are a bit messed up.

**Comment by Arash-Afshar** [comment]
Merging to unblock the next PR. Will come back and revisit the `unwrap` and regex issues in another iteration.

**Pull request merged**